Homework 1

Arthasmethil opened the pull request:
Homework 1, pull request

Arthasmethil committed:
Homework 1

shoggy commented on home-tasks/pom.xml:
```diff
@@ -12,14 +12,114 @@
+            <version>${assertj-core.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.epam.calculator</groupId>
+            <artifactId>epam-calculator</artifactId>
+            <version>${epam-calculator.version}</version>
+        </dependency>
```
Which actions were performed to put this artifact into local repository?

Arthasmethil replied:
I used a command for windows console: 
mvn install:install-file -Dfile=C:\Users\artha\Downloads\HW1_Frameworks\calculator-1.0.jar -DgroupId=com.epam.calculator -DartifactId=epam-calculator -Dversion=1.0 -Dpackaging=jar -DgeneratePom=true

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw1/functions/AdditionOperationTestDouble.java:
```diff
@@ -0,0 +1,31 @@
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class AdditionOperationTestDouble {
+    Calculator calculator;
+
+    @BeforeMethod
+    public void setUp() {
+        calculator = new Calculator();
+    }
```
These test classes have a lot in common. At least they test calculator's operations. So would be better to make them extend some parent class, where test configuration occurs (before/after methods)

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw1/functions/AdditionOperationTestDouble.java:
```diff
@@ -0,0 +1,31 @@
+public class AdditionOperationTestDouble {
+    Calculator calculator;
+
+    @BeforeMethod
```
FYI, wisely choose method/class/... for configurational methods. Some actions should be performed for each test-method call, some, like initialization of SUT (system under test) should happen once per suite.

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw1/functions/DivideByZeroOperationTest.java:
```diff
@@ -0,0 +1,31 @@
+        calculator = new Calculator();
+    }
+
+    @Test(
+        expectedExceptions = NumberFormatException.class,
+        dataProviderClass = DataProviderForCalculator.class,
+        dataProvider = "getTestDataForDivideByZero")
+    public void divideTest(long a, long b) throws NumberFormatException {
+        assertThat(calculator.div(a, b)).isEqualTo(a / b);
```
I think this method can be moved to `DivideOperationTest.java`

shoggy replied:
Division in `.isEqualTo` is not a good idea here, because it obviously can produce ArithmeticException
Please read this one: https://www.baeldung.com/assertj-exception-assertion

shoggy commented on home-tasks/src/test/resources/suites/additionAndSubtractOperations.xml:
```diff
@@ -0,0 +1,12 @@
+
+<suite name="SimpleFunctions" verbose="1" >
+  <test name="AdditionAndSubtractTest">
+    <classes>
+      <class name="com.epam.tc.hw1.functions.SubtractOperationTestDouble"/>
+      <class name="com.epam.tc.hw1.functions.SubtractOperationTestLong"/>
+      <class name="com.epam.tc.hw1.functions.AdditionOperationTestDouble"/>
+      <class name="com.epam.tc.hw1.functions.AdditionOperationTestLong"/>
+    </classes>
```
Try to use class level `groups` instead of explicit declarations
http://makeseleniumeasy.com/2018/07/03/testng-tutorials-30-groups-attribute-at-class-level-in-testng/

shoggy commented on home-tasks/pom.xml:
```diff
@@ -12,14 +12,114 @@
+        <assertj-core.version>3.22.0</assertj-core.version>
+        <epam-calculator.version>1.0</epam-calculator.version>
+        <slf4j-simple.version>1.7.35</slf4j-simple.version>
+        <testng.simpleSuite.config.path>src/test/resources/suites</testng.simpleSuite.config.path>
```
As improvement 
```suggestion
        <testng.simpleSuite.config.path>${project.build.testOutputDirectory}/suites</testng.simpleSuite.config.path>
```

shoggy reviewed (changes requested):
## Test structure requirements:
- [x] `.gitignore` includes `target, .iml, .idea, ...`
- [x] Maven project
- [x] POM xml has following sections:
  - [x] artifact attributes
   - `artifactId` should be in form `firstname_lastname` as described in the presentation from introduction (Dmitrii Lebedev)
  - [x] properties
  - [x] dependencies
  - [x] plugins
  - [x] profiles
- [x] `calculator-1.0.jar` is added as maven dependency
- [x] TestNG is used as a testing framework
- [x] For test maven build maven-surefire-plugin is used
- [x] Code conforms to the Java Code Convention
- [x] `maven-checkstyle-plugin` is configured to use specific rules and is run during  `validate` phase

## File placement requirements:
- [x] Classes should be placed into `src/test/java/com/epam/tc/hw1` package and sub-packages
- [x] Homework should be developed in new Java classes
- [x] Test methods' names should reflect checking functionality

## Tests requirements:
- [x] Checkstyle plugin should not produce errors
- [x] You should test several Calculator operations (at least 4)
- [x] Test for each operation should be placed in separate classes
- [x] Create a test suite, which could run all tests
- [x] Split test for the two groups and create a suite for each group:
  - [x] Tests for add and subtract operations
  - [x] Tests for multiply and divide operations
- [x] All test data should be provided to test through the DataProvider
- [x] Profiles
  - [x] Create 3 maven profiles which allow choosing test suite
  - [x] by default should work profile which runs add and subtract operations tests
- [x] Test must be run using maven
- [x] AssertJ is used instead of basic TestNg asserts

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw1/functions/AdditionOperationTestDouble.java:
```diff
@@ -0,0 +1,31 @@
+        dataProvider = "getTestDataForAdditionWithDouble")
+    public void additionTestWithDouble(double a, double b, double expected) {
+        double actual =  calculator.sum(a, b);
+        assertThat(actual).isEqualTo(expected);
```
Floating point values should be compared using delta 
http://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/FloatingPointNumberAssert.html

Arthasmethil committed:
Scope of work:
-AbstractCalculatorOperationTest was created
-Delta was added in tests with double and float data
-Class DivideByZeroOperationTest was combined with DivideOperationTest
-Test suites class level was changed from 'classes' to 'group'
-refactor POM, test with exception

Arthasmethil committed:
Create .gitignore

Additional .gitignore for main directoty

shoggy reviewed (approved):
Good job!
